diff -rupN BeamformIt-3.51.org/CMakeLists.txt BeamformIt-3.51.changed/CMakeLists.txt
--- BeamformIt-3.51.org/CMakeLists.txt	2014-06-25 10:30:13.000000000 -0400
+++ BeamformIt-3.51.changed/CMakeLists.txt	2015-10-16 10:28:19.333848099 -0400
@@ -20,7 +20,7 @@ set (CMAKE_C_FLAGS_RELEASE         "-D_R
 set (CMAKE_CXX_FLAGS_RELEASE       "-DLINUX -Wall -W -O3 -DNDEBUG -Wno-unknown-pragmas -Wno-unused-result -Wno-unused-variable")
 
 ##find libraries installed in the system
-FIND_LIBRARY(LIBSNDFILE sndfile)
+FIND_LIBRARY(LIBSNDFILE sndfile ../libsndfile-1.0.25/lib/)
 
 ##set particular settings for each architecture
 set (FFTREAL_HDRS ${CMAKE_CURRENT_SOURCE_DIR}/external/FFTReal-2.11)
@@ -37,7 +37,7 @@ endif()
 ##set some generic link directories
 LINK_DIRECTORIES(${os_link_libs})
 set (CMAKE_INCLUDE_SYSTEM_FLAG_CXX "-isystem") #needed for the SYSTEM link directories to work on osx
-INCLUDE_DIRECTORIES(SYSTEM ${os_include_libs})
+INCLUDE_DIRECTORIES(SYSTEM ${os_include_libs} ../libsndfile-1.0.25/include/)
 INCLUDE_DIRECTORIES(${CMAKE_CURRENT_SOURCE_DIR}/includes)
 
 #gather project files (note that when adding a new file cmake needs to be rerun)
diff -rupN BeamformIt-3.51.org/src/delaysum.cc BeamformIt-3.51.changed/src/delaysum.cc
--- BeamformIt-3.51.org/src/delaysum.cc	2014-06-25 12:35:44.000000000 -0400
+++ BeamformIt-3.51.changed/src/delaysum.cc	2015-10-19 13:11:13.062184285 -0400
@@ -12,7 +12,7 @@
 #include <iostream>
 #include <cstring>
 
-#include "samplerate.h"
+//#include "samplerate.h"
 #include "sndfile.h"
 #include "global.h"
 #include "delaysum.h"
@@ -423,7 +423,8 @@ void DelaySum::compute_scalling_factor()
   float max_val_segment[m_numCh];
   vector<float> median_val_segment;
   int num_segments=0;
-  int total_num_segments = 10; //we will only compute (at max) these number of segments
+  int total_num_segments_initial = 10; //we will only compute (at max) these number of segments
+  int total_num_segments = total_num_segments_initial; // it can be variable
   int signal_scroll; //advancement in the signal for the next segment
   int segment_duration = 10 * m_sampleRate; //10 seconds segments
   float sum_weighting;
@@ -442,9 +443,16 @@ void DelaySum::compute_scalling_factor()
     num_segments = 0;
     start_place = (long)(m_UEMGap);
     //check if we have enough signal for the number of segments
-    if(((m_frames - start_place)/segment_duration) < total_num_segments)
+    if(((m_frames - start_place)/segment_duration) < total_num_segments_initial)
     {
-      total_num_segments = int((m_frames - start_place)/segment_duration); 
+      total_num_segments = int((m_frames - start_place) / segment_duration);
+      if (total_num_segments < 1)
+      {
+				total_num_segments = 1;
+				segment_duration = m_frames - start_place; //we do not have more frames
+      }
+      char tmp_string[1024]; sprintf(tmp_string,"Set the total number of segments from %d to %d for computing scalling factor, with segment duration %d\n", total_num_segments_initial, total_num_segments, segment_duration);
+      m_log.print_this(tmp_string, 1);
     }
     signal_scroll = int((m_frames - start_place)/total_num_segments);
     char tmp_string[1024]; sprintf(tmp_string,"Processing channel %d\n", channel_count);
diff -rupN BeamformIt-3.51.org/src/fileinout.cc BeamformIt-3.51.changed/src/fileinout.cc
--- BeamformIt-3.51.org/src/fileinout.cc	2014-06-25 09:27:05.000000000 -0400
+++ BeamformIt-3.51.changed/src/fileinout.cc	2015-10-16 10:28:19.333848099 -0400
@@ -1,7 +1,7 @@
 #include <stdlib.h>
 #include <cstring>
 #include "fileinout.h"
-#include "samplerate.h" 
+//#include "samplerate.h" 
 
 //char tmp_string[1024];
 
diff -rupN BeamformIt-3.51.org/src/tdoa.cc BeamformIt-3.51.changed/src/tdoa.cc
--- BeamformIt-3.51.org/src/tdoa.cc	2014-06-25 10:55:40.000000000 -0400
+++ BeamformIt-3.51.changed/src/tdoa.cc	2015-10-16 10:28:19.333848099 -0400
@@ -10,7 +10,7 @@
 #include "tdoa.h"
 #include "fileinout.h"
 #include "global.h"
-#include "fftw3.h"
+//#include "fftw3.h"
 
 //char tmp_string[1024];
 
