--- a/Pre-built.2/include/pthread.h    2016-02-14 13:49:16.027430500 -0500
+++ b/Pre-built.2/include/pthread_new.h    2016-02-14 13:53:40.729116100 -0500
@@ -317,10 +317,12 @@
 #define HAVE_STRUCT_TIMESPEC
 #if !defined(_TIMESPEC_DEFINED)
 #define _TIMESPEC_DEFINED
+#if defined(_MSC_VER) && _MSC_VER >= 1900 && defined(_CRT_NO_TIME_T)
 struct timespec {
         time_t tv_sec;
         long tv_nsec;
 };
+#endif /* _CRT_NO_TIME_T */
 #endif /* _TIMESPEC_DEFINED */
 #endif /* HAVE_STRUCT_TIMESPEC */
