--- src/mdiadapt.cpp	2014-03-05 00:43:33.000000000 -0800
+++ src/mdiadapt.cpp.new	2014-03-05 00:42:41.000000000 -0800
@@ -20,6 +20,7 @@
 
 #include <cmath>
 #include <string>
+#include <vector>
 #include <assert.h>
 #include "mfstream.h"
 #include "mempool.h"
@@ -1112,7 +1113,7 @@
 	}else{
 		cerr << "savebin: " << filename << "\n";		
 	}
-  streampos pos[lmsize()+1];
+  vector<streampos> pos(lmsize()+1);
   int num[lmsize()+1];
 	
 	int maxlev=lmsize();
@@ -1303,7 +1304,7 @@
   fstream out(filename,ios::out);
 	//  out.precision(15);  
 	
-  streampos pos[lmsize()+1];
+  vector<streampos> pos(lmsize()+1);
   int num[lmsize()+1];
   char buff[100];
 
--- src/interpolate-lm.cpp	2014-03-05 00:43:37.000000000 -0800
+++ src/interpolate-lm.cpp.new	2014-03-05 00:43:01.000000000 -0800
@@ -258,7 +258,7 @@
   //Learning mixture weights
   if (learn){
 	
-    std::vector<float> p[N]; //LM probabilities
+    std::vector< std::vector<float> > p(N); //LM probabilities
     float c[N]; //expected counts
     float den,norm; //inner denominator, normalization term
     float variation=1.0; // global variation between new old params
